presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -975,12 +974,9 @@ else if (criteria instanceof JoinOn) {
 
                 Analyzer.verifyNoAggregatesOrWindowFunctions(metadata.getFunctionRegistry(), expression, "JOIN clause");
 
-                Expression canonicalized = CanonicalizeExpressions.canonicalizeExpression(expression);
-                analyzeExpression(canonicalized, output);
-
                 Set<Expression> postJoinConjuncts = new HashSet<>();
 
-                for (Expression conjunct : ExpressionUtils.extractConjuncts(canonicalized)) {
+                for (Expression conjunct : ExpressionUtils.extractConjuncts(expression)) {
                     conjunct = ExpressionUtils.normalize(conjunct);
```

Comment:
@sopel39: `analysis.recordSubqueries(node, analyzeExpression(postJoinPredicate, output));
` does the analysis of the conjucts, so the coercions for normalized conjucts will be registered in `Analysis`.

I believe we could clean up extracting conjucts as well but I think it's a separate issue because current approach is already correct and solves a problem.